File: src/cn/rieon/idea/vim/AutoSwitchComponent.java
package cn.rieon.idea.vim;

import com.intellij.ide.FrameStateListener;
import com.intellij.ide.FrameStateManager;
import com.intellij.notification.Notification;
import com.intellij.notification.NotificationType;
import com.intellij.notification.Notifications;
import com.intellij.openapi.command.CommandAdapter;
import com.intellij.openapi.command.CommandEvent;
import com.intellij.openapi.command.CommandListener;
import com.intellij.openapi.command.CommandProcessor;
import com.intellij.openapi.components.ApplicationComponent;
import com.intellij.openapi.diagnostic.Logger;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * @author Rieon Ke <rieon@rieon.cn>
 * @version 1.0.0
 * @since 2017/5/20
 */
public class AutoSwitchComponent implements ApplicationComponent {

    private static final Logger LOG = Logger.getInstance(AutoSwitchComponent.class);
    /**
     * last input source
     */
    private String lastInputSource = null;

    private Boolean inNormal = true;

    public static final String defaultEnglishInputSource = "com.apple.keylayout.ABC";

    /**
     * switch input source in these vim states
     */
    private static final List<String> VimStateSwitchToABCOnly = new ArrayList<String>(){{
        // to visual mode
        add("Vim Exit Visual Mode");
        add("Vim Toggle Line Selection");
        add("Vim Toggle Block Selection");
        add("Vim Toggle Character Selection");
        // to normal mode
        add("Vim Exit Insert Mode");
        //EditorChange

    }};

    /**
     * switch to last input source in these vim states
     */
    private static final List<String> VimStateSwitchToLastIm = new ArrayList<String>(){{

        add("Vim Enter");
        add("Vim Insert at Line Start");
        add("Vim Insert New Line Above");
        add("Vim Insert New Line Below");
        add("Vim Insert Previous Text");
        add("Vim Insert Previous Text");
        add("Vim Insert Register");
        add("Vim Toggle Insert/Replace");
        add("Vim Change Line");
        add("Vim Change Character");
        add("Vim Change Characters");
        add("Vim Replace");
        add("Vim Insert After Cursor");
        add("Vim Insert After Line End");
        add("Vim Insert Before Cursor");
        add("Vim Insert Before First non-Blank");
        add("Vim Insert Character Above Cursor");
        add("Vim Insert Character Below Cursor");
        add("Vim Delete Inserted Text");
        add("Vim Delete Previous Word");

    }};

    public AutoSwitchComponent() {
    }

    /**
     * init component and add listener
     */
    @Override
    public void initComponent() {

        LOG.info("INIT COMPONENT");

        lastInputSource = getCurrentInputSourceId();
        CommandProcessor.getInstance().addCommandListener(getCommandListener());
        FrameStateManager.getInstance().addListener(getFrameStateListener());

    }

    /**
     * auto switch input source while vim state changed
     * @return CommandListener
     */
    private CommandListener getCommandListener(){
        return new CommandAdapter() {
            @Override
            public void beforeCommandFinished(CommandEvent event) {

                String commandName = event.getCommandName();

                if (VimStateSwitchToABCOnly.contains(commandName)){
                    lastInputSource = getCurrentInputSourceId();
                    inNormal = true;
                    if(lastInputSource == null || Objects.equals(lastInputSource, defaultEnglishInputSource))
                        return;
                    switchTo(defaultEnglishInputSource);
                }else if (VimStateSwitchToLastIm.contains(commandName)){
                    String current = getCurrentInputSourceId();
                    inNormal = false;
                    if (current == null || current.equals(lastInputSource))
                        return;
                    switchTo(lastInputSource);
                }else if ("Typing".equals(commandName)){
                    inNormal = false;
                }else if ("".equals(commandName)){
                    if (inNormal)
                        switchTo(defaultEnglishInputSource);
                }
            }
        };

    }

    /**
     * restore last input source while idea re-focused
     * @return FrameStateListener
     */
    private FrameStateListener getFrameStateListener(){

        return new FrameStateListener.Adapter() {
            @Override
            public void onFrameDeactivated() {
                lastInputSource = getCurrentInputSourceId();
            }

            @Override
            public void onFrameActivated() {
                String current = getCurrentInputSourceId();
                if (current == null || current.equals(lastInputSource))
                    return;
                switchTo(lastInputSource);
            }
        };
    }

    /**
     * get current input source
     * @return String currentInputSource
     */
    private String getCurrentInputSourceId(){

        String current = InputSource.getCurrentInputSource();

        LOG.info("CURRENT INPUT SOURCE  " + current);
        if (current == null || current.length() <= 0){
            Notification notification = new Notification("Switch IME Error","Switch IME Error","Get current input source faild",NotificationType.ERROR);
            Notifications.Bus.notify(notification);
            LOG.info("GET CURRENT INPUT SOURCE FAILED");
        }

        return current;

    }

    void switchTo(String source){

        LOG.info("SWITCH TO INPUT SOURCE  " + source);
        if (source == null)
            return;
        if (!InputSource.switchTo(source,InputSource.BY_ID)) {
            Notification notification = new Notification("Switch IME Error", "Switch IME Error", "Switch IME Failed", NotificationType.ERROR);
            Notifications.Bus.notify(notification);
            LOG.info("SWITCH TO INPUT SOURCE FAILED");
        }
    }

    @Override
    public void disposeComponent() {
        // TODO: insert component disposal logic here
    }

    @Override
    @NotNull
    public String getComponentName() {
        return "AutoSwitchComponent";
    }

}

File: src/main/java/cn/rieon/idea/plugin/AutoSwitchIm/component/AutoSwitchComponent.kt
package cn.rieon.idea.plugin.AutoSwitchIm.component

import cn.rieon.idea.plugin.AutoSwitchIm.Constraints
import cn.rieon.idea.plugin.AutoSwitchIm.provider.ConfigurationProvider
import cn.rieon.idea.plugin.AutoSwitchIm.util.InputSourceUtil
import com.intellij.ide.FrameStateListener
import com.intellij.ide.FrameStateManager
import com.intellij.notification.Notification
import com.intellij.notification.NotificationType
import com.intellij.notification.Notifications
import com.intellij.openapi.command.CommandAdapter
import com.intellij.openapi.command.CommandEvent
import com.intellij.openapi.command.CommandListener
import com.intellij.openapi.command.CommandProcessor
import com.intellij.openapi.components.ApplicationComponent
import com.intellij.openapi.diagnostic.Logger
import java.util.*

/**
 * @author Rieon Ke <rieon></rieon>@rieon.cn>
 * *
 * @version 1.0.0
 * *
 * @since 2017/5/20
 */
class AutoSwitchComponent : ApplicationComponent {

    private var currentInputSource: String? = null

    /**
     * last input source
     */
    private var lastOutVimInsert: String? = null
    private var lastInVimInsert: String? = null
    private var lastOutIdea: String? = null
    private var lastInIdea: String? = null

    private var inNormal: Boolean? = true

    private var configurationProvider: ConfigurationProvider? = null



    /**
     * init component and add listener
     */
    override fun initComponent() {

        LOG.info("INIT COMPONENT")

        currentInputSource = currentInputSourceId
        CommandProcessor.getInstance().addCommandListener(commandListener)
        FrameStateManager.getInstance().addListener(frameStateListener)

        /**
         * Reload configuration from configuration provider
         */

        configurationProvider = ConfigurationProvider.instance

        if (configurationProvider!!.outVimInsertConfig != null) {

            outVimInsertConfig = configurationProvider!!.outVimInsertConfig!!

            LOG.info("USE CONFIG INPUT SOURCE FOR OUT VIM INSERT MODE" + outVimInsertConfig)
        }

        if (configurationProvider!!.InVimInsertConfig != null) {

            inVimInsertConfig = configurationProvider!!.InVimInsertConfig!!

            LOG.info("USE CONFIG INPUT SOURCE FOR In VIM INSERT MODE" + inVimInsertConfig)
        }
        if (configurationProvider!!.outOfIdeaConfig != null) {

            outOfIdeaConfig = configurationProvider!!.outOfIdeaConfig!!

            LOG.info("USE CONFIG INPUT SOURCE FOR OUT OF IDEA" + outOfIdeaConfig)
        }
        if (configurationProvider!!.inIdeaConfig != null) {

            inIdeaConfig = configurationProvider!!.inIdeaConfig!!

            LOG.info("USE CONFIG INPUT SOURCE FOR IDEA FOCUSED" + inIdeaConfig)
        } //load complete

        LOG.info("INIT SUCCESSFUL")
    }

    /**
     * auto switch input source while plugin state changed
     * @return CommandListener
     */
    private val commandListener: CommandListener
        get() = object : CommandAdapter() {
            override fun beforeCommandFinished(event: CommandEvent?) {

                val commandName = event!!.commandName

                LOG.info("Current Command Name is " + commandName)

                if (commandName == null || commandName == "null"){
                    return;
                }

                /**
                 * out of vim insert mode
                 */
                if (OUT_VIM_INSERT_MODE.contains(commandName)) {
                    handleOutVimInsertSwitch()
                    /**
                     * into vim insert mode
                     */
                } else if (IN_VIM_INSERT_MODE.contains(commandName)) {
                    handleInVimInsertSwitch()

                    /**
                     * start typing => not in vim normal mode
                     */
                } else if ("Typing" == commandName) {


                    // set current vim mode to normal
                    inNormal = false

                } else if ("" == commandName) {

                    // if current in normal mode
                    if (inNormal!!)
                        handleOutVimInsertSwitch()

                }
            }
        }

    /**
     * restore last input source while idea re-focused
     * @return FrameStateListener
     */
    private val frameStateListener: FrameStateListener
        get() = object : FrameStateListener.Adapter() {
            override fun onFrameDeactivated() {
                handleOutIdeaSwitch()
            }

            override fun onFrameActivated() {
                handInIdeaSwitch()
            }
        }

    /**
     * get current input source
     * @return String currentInputSource
     */
    private val currentInputSourceId: String?
        get() {

            val current = InputSourceUtil.currentInputSource

            LOG.info("CURRENT INPUT SOURCE  " + current)
            if (current.isEmpty()) {
                val notification = Notification("Switch IME Error", "Switch IME Error", "Get current input source faild", NotificationType.ERROR)
                Notifications.Bus.notify(notification)
                LOG.info("GET CURRENT INPUT SOURCE FAILED")
            }

            return current

        }

    internal fun switchTo(source: String?) {

        LOG.info("SWITCH TO INPUT SOURCE  " + source!!)
        if (!InputSourceUtil.switchTo(source)) {
            val notification = Notification("Switch IME Error", "Switch IME Error", "Switch IME Failed", NotificationType.ERROR)
            Notifications.Bus.notify(notification)
            LOG.info("SWITCH TO INPUT SOURCE FAILED")
        }
    }

    override fun disposeComponent() {
        // TODO: insert component disposal logic here
    }

    override fun getComponentName(): String {
        return "AutoSwitchComponent"
    }

    companion object {

        private val LOG = Logger.getInstance(AutoSwitchComponent::class.java)

        var outVimInsertConfig = Constraints.CONFIG_INPUT_DEFAULT
        var inVimInsertConfig = Constraints.CONFIG_INPUT_DEFAULT
        var inIdeaConfig = Constraints.CONFIG_INPUT_DEFAULT
        var outOfIdeaConfig = Constraints.CONFIG_INPUT_DEFAULT


        /**
         * switch input source in these plugin states
         */
        private val OUT_VIM_INSERT_MODE = Arrays.asList(
                "Vim Exit Visual Mode",
                "Vim Toggle Line Selection",
                "Vim Toggle Block Selection",
                "Vim Toggle Character Selection",
                "Vim Exit Insert Mode")

        /**
         * switch to last input source in these plugin states
         */
        private val IN_VIM_INSERT_MODE = Arrays.asList(
                "Vim Enter",
                "Vim Insert at Line Start",
                "Vim Insert New Line Above",
                "Vim Insert New Line Below",
                "Vim Insert Previous Text",
                "Vim Insert Previous Text",
                "Vim Insert Register",
                "Vim Toggle Insert/Replace",
                "Vim Change Line",
                "Vim Change Character",
                "Vim Change Characters",
                "Vim Replace",
                "Vim Insert After Cursor",
                "Vim Insert After Line End",
                "Vim Insert Before Cursor",
                "Vim Insert Before First non-Blank",
                "Vim Insert Character Above Cursor",
                "Vim Insert Character Below Cursor",
                "Vim Delete Inserted Text",
                "Vim Delete Previous Word")
    }

    internal fun handleInVimInsertSwitch(){

        inNormal = false
        if (inVimInsertConfig == Constraints.CONFIG_INPUT_NONE) return;

        //get current input source
        currentInputSource = currentInputSourceId
        //record last input source state for last-in-vim-insert-mode
        lastOutVimInsert = currentInputSource

        handle(currentInputSource,lastInVimInsert, inVimInsertConfig)

    }

    internal fun handleOutVimInsertSwitch(){

        inNormal = true
        if (outVimInsertConfig == Constraints.CONFIG_INPUT_NONE) return;
        //get current input source
        currentInputSource = currentInputSourceId
        //record last input source state for last-in-vim-insert-mode
        lastInVimInsert = currentInputSource


        handle(currentInputSource,lastOutVimInsert, outVimInsertConfig)

    }

    internal fun handleOutIdeaSwitch(){

        if (outOfIdeaConfig == Constraints.CONFIG_INPUT_NONE) return;
        //get current input source
        currentInputSource = currentInputSourceId
        //record last input source state for last-in-vim-insert-mode
        lastInIdea = currentInputSource

        handle(currentInputSource,lastOutIdea, outOfIdeaConfig)

    }

    internal fun handInIdeaSwitch(){

        if (inIdeaConfig == Constraints.CONFIG_INPUT_NONE) return;
        //get current input source
        currentInputSource = currentInputSourceId
        //record last input source state for last-in-vim-insert-mode
        lastOutIdea = currentInputSource

        handle(currentInputSourceId,lastInIdea, inIdeaConfig)

    }

    internal fun handle(current:String?,last:String?,config:String) {

        if (current == null || current == config) return

        when (config) {

            Constraints.CONFIG_INPUT_LAST -> {

                if (last == null || current == last) {
                    return
                } else {
                    switchTo(last)
                }

            }

            Constraints.CONFIG_INPUT_NONE -> return

        // use input source in configuration
            else -> {
                switchTo(config)
            }
        }
    }
}
